store/engine/embedded/raft.go

```diff
@@ -0,0 +1,550 @@
+	 }
+ 
+	 rc.transport = &rafthttp.Transport{
+		 Logger:      logger.Get(),
+		 ID:          types.ID(rc.id),
+		 ClusterID:   0x1000,
+		 Raft:        rc,
+		 ServerStats: stats.NewServerStats("", ""),
+		 LeaderStats: stats.NewLeaderStats(zap.NewExample(), strconv.Itoa(rc.id)),
+		 ErrorC:      make(chan error),
+ 
+		 DialRetryFrequency: 1,
+	 }
+ 
+	 if err := rc.transport.Start(); err != nil {
+		 logger.Get().With(zap.Error(err)).Panic("Failed to start raft http server")
+	 }
+	 for i := range rc.peers {
+		 if i+1 != rc.id {
+			 rc.transport.AddPeer(types.ID(i+1), []string{rc.peers[i]})
+		 }
+	 }
+ 
+	 go rc.serveRaft()
+	 go rc.serveChannels()
+ }
+ 
+ // stop closes http, closes all channels, and stops raft.
+ func (rc *raftNode) stop() {
+	 rc.stopHTTP()
+	 close(rc.commitC)
+	 close(rc.errorC)
+	 rc.node.Stop()
+ }
+ 
+ func (rc *raftNode) stopHTTP() {
+	 rc.transport.Stop()
+	 close(rc.httpStopCh)
+	 <-rc.httpDoneCh
+ }
+ 
+ func (rc *raftNode) publishSnapshot(snapshotToSave raftpb.Snapshot) {
+	 if raft.IsEmptySnap(snapshotToSave) {
+		 return
+	 }
+ 
+	 logger.Get().Sugar().Infof("Publishing snapshot at index %d", rc.snapshotIndex)
+	 defer logger.Get().Sugar().Infof("Finished publishing snapshot at index %d", rc.snapshotIndex)
+ 
+	 if snapshotToSave.Metadata.Index <= rc.appliedIndex {
+		 logger.Get().Sugar().Fatalf("snapshot index [%d] should > progress.appliedIndex [%d]", snapshotToSave.Metadata.Index, rc.appliedIndex)
+	 }
+	 rc.commitC <- nil // trigger kvstore to load snapshot
+ 
+	 rc.confState = snapshotToSave.Metadata.ConfState
+	 rc.snapshotIndex = snapshotToSave.Metadata.Index
+	 rc.appliedIndex = snapshotToSave.Metadata.Index
+ }
+ 
+ var snapshotCatchUpEntriesN uint64 = 10000
+ 
+ func (rc *raftNode) maybeTriggerSnapshot(applyDoneC <-chan struct{}) {
+	 if rc.appliedIndex-rc.snapshotIndex <= rc.snapCount {
+		 return
+	 }
+ 
+	 // wait until all committed entries are applied (or server is closed)
+	 if applyDoneC != nil {
+		 select {
+		 case <-applyDoneC:
+		 case <-rc.stopCh:
+			 return
+		 }
+	 }
+ 
+	 logger.Get().Sugar().Infof("start snapshot [applied index: %d | last snapshot index: %d]", rc.appliedIndex, rc.snapshotIndex)
+	 data, err := rc.getSnapshot()
+	 if err != nil {
+		 logger.Get().With(zap.Error(err)).Panic("Failed to snapshot")
+	 }
+	 snapshot, err := rc.raftStorage.CreateSnapshot(rc.appliedIndex, &rc.confState, data)
+	 if err != nil {
+		 panic(err)
+	 }
+	 if err := rc.saveSnap(snapshot); err != nil {
+		 panic(err)
+	 }
+ 
+	 compactIndex := uint64(1)
+	 if rc.appliedIndex > snapshotCatchUpEntriesN {
+		 compactIndex = rc.appliedIndex - snapshotCatchUpEntriesN
+	 }
+	 if err := rc.raftStorage.Compact(compactIndex); err != nil {
+		 if !errors.Is(err, raft.ErrCompacted) {
+			 panic(err)
+		 }
+	 } else {
+		 logger.Get().Sugar().Infof("compacted log at index %d", compactIndex)
+	 }
+ 
+	 rc.snapshotIndex = rc.appliedIndex
+ }
+ 
+ func (rc *raftNode) serveChannels() {
```

Comment:
What's the meaning of Channels here? would you mind refactoring this function to make it shorter by moving the routine into a single function?

Reply:
`serveRaft` is to setup TCP server to communicate with peers, while `serveChannels` listen on inbound and outbound channels to do actual business.